src/DeviceInput/InputDevices/deviceSourceManager.ts

```diff
@@ -0,0 +1,308 @@
+     */
+    public onAfterDeviceDisconnectedObservable = new Observable<{ deviceType: DeviceType, deviceSlot: number }>();
+
+    // Private Members
+    private _devices: Array<Array<DeviceSource<DeviceType>>>;
+    private _firstDevice: Array<number>;
+    private _deviceInputSystem: DeviceInputSystem;
+
+    /**
+     * Default Constructor
+     * @param engine engine to pull input element from
+     * @param enableObserveEvents boolean to enable use of observe events
+     */
+    constructor(engine: Engine, enableObserveEvents: boolean = false) {
+        const numberOfDeviceTypes = Object.keys(DeviceType).length / 2;
+        this._devices = new Array<Array<DeviceSource<DeviceType>>>(numberOfDeviceTypes);
+        this._firstDevice = new Array<number>(numberOfDeviceTypes);
+        this._deviceInputSystem = new DeviceInputSystem(engine);
+
+        this._deviceInputSystem.onDeviceConnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._addDevice(deviceName);
+            this.onAfterDeviceConnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+        this._deviceInputSystem.onDeviceDisconnected = (deviceName) => {
+            const deviceType = this._getDeviceTypeFromName(deviceName);
+            const deviceSlot = this._getDeviceSlot(deviceName);
+
+            this.onBeforeDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+            this._removeDevice(deviceName);
+            this.onAfterDeviceDisconnectedObservable.notifyObservers({ deviceType, deviceSlot });
+        };
+
+        if (!this._deviceInputSystem.observeInput && enableObserveEvents) {
+            this._deviceInputSystem.observeInput = (deviceName, inputIndex, previousState, currentState) => {
+                const deviceType = this._getDeviceTypeFromName(deviceName);
+                const deviceSlot = this._getDeviceSlot(deviceName);
+
+                if (deviceType == DeviceType.Keyboard || deviceType == DeviceType.Mouse || deviceType == DeviceType.Touch) {
+                    this.getDeviceSource(deviceType, deviceSlot)?.onInputChangedObservable.notifyObservers({ inputIndex, previousState, currentState });
+                }
+            };
+        }
+    }
+
+    // Public Functions
+    /**
+     * Checks for current device input value, given DeviceType, slot, and inputIndex
+     * @param type Enum specifiying device type
+     * @param inputIndex Index of device input
+     * @param deviceSlot "Slot" or index that device is referenced in
+     * @returns Current value of input
+     */
+    public getInput<T extends DeviceType>(type: T, inputIndex: DeviceInputs<T>, deviceSlot: number = this._firstDevice[type]): Nullable<number> {

```

Comment:
Is this function still useful? It is duplicate functionality with getDeviceSource and then calling getInput?

Reply:
Yea I find this rather confusing. I would vote to remove it to simplify the API.

Reply:
I originally wanted to leave this to keep with the initial idea of simply asking the DSM for an input but using the getDeviceSource function is functionally the same thing.  Removed.

Reply:
So I was testing the removal of this and it breaks my original ideal of single line poll-and-forget.  Before, you could just keep asking and either get nothing or get input.  Now, it will fail if the device you're listening to hasn't been added yet.  This can be worked around by just checking for the device before polling.  I do want to revisit this but for now I can leave it out.

Reply:
I'm not sure I understand. You can use the other function to do it all in one line no?

Reply:
You can, if a DeviceSource exists for it.  Otherwise, you're trying to get input from a null object.

Reply:
With newer versions of TypeScript, it is easy to do this in one-line:
```javascript
deviceSourceManager.getDeviceSource(...)?.getInput(...)
```

Reply:
That works for TypeScript but it looks like it isn't available yet for JavaScript (proposal for it is finished though so it could be coming soon.).